js/cards/disk2.ts

```diff
@@ -283,71 +253,74 @@ interface State {
 }
 
 function getDriveState(drive: Drive): DriveState {
-    const result: DriveState = {
-        format: drive.format,
-        encoding: drive.encoding,
-        volume: drive.volume,
-        tracks: [],
+    return {
+        disk: getDiskState(drive.disk),
+        metadata: {...drive.metadata},
+        readOnly: drive.readOnly,
         track: drive.track,
         head: drive.head,
         phase: drive.phase,
-        readOnly: drive.readOnly,
         dirty: drive.dirty,
-        trackMap: [],
-        rawTracks: [],
-        metadata: { ...drive.metadata },
     };
-
-    if (isNibbleDrive(drive)) {
-        for (let idx = 0; idx < drive.tracks.length; idx++) {
-            result.tracks.push(new Uint8Array(drive.tracks[idx]));
-        }
-    }
-    if (isWozDrive(drive)) {
-        result.trackMap = [...drive.trackMap];
-        for (let idx = 0; idx < drive.rawTracks.length; idx++) {
-            result.rawTracks.push(new Uint8Array(drive.rawTracks[idx]));
-        }
-    }
-    return result;
 }
 
-function setDriveState(state: DriveState) {
-    let result: Drive;
-    if (state.encoding === ENCODING_NIBBLE) {
-        result = {
-            format: state.format,
-            encoding: ENCODING_NIBBLE,
-            volume: state.volume,
+function getDiskState(disk: NoFloppyDisk): NoFloppyDisk;
+function getDiskState(disk: NibbleDisk): NibbleDisk;
+function getDiskState(disk: WozDisk): WozDisk;
+function getDiskState(disk: FloppyDisk): FloppyDisk;
+function getDiskState(disk: FloppyDisk): FloppyDisk {
+    if (isNoFloppyDisk(disk)) {
+        return {
+            encoding: disk.encoding,
+            metadata: {...disk.metadata},
+            readOnly: disk.readOnly,
+        };
+    }
+    if (isNibbleDisk(disk)) {
+        const result: NibbleDisk = {
+            format: disk.format,
+            encoding: disk.encoding,
+            volume: disk.volume,
             tracks: [],
-            track: state.track,
-            head: state.head,
-            phase: state.phase,
-            readOnly: state.readOnly,
-            dirty: state.dirty,
-            metadata: { ...state.metadata },
+            readOnly: disk.readOnly,
+            metadata: { ...disk.metadata },
```

Comment:
While you're in here, you might consider destructuring all these values in advance so you can take advantage of object shorthand.

Reply:
If by "destructuring" you mean including all of the metadata fields in the same record as the drive state, then I would really prefer not to do that. This whole adventure has been centered around getting the disk fields _out_ of the drive structure and into its own sub-structure. From my understanding, the destructuring syntax doesn't really work in this case because objects are copied by reference. (Maybe a judicious use of [`structuredClone`](https://developer.mozilla.org/en-US/docs/Web/API/structuredClone) or `JSON.stringify` would be sufficient?)

This brings up the important question, though, of how to think about the VM state.

Originally, the state used entirely separate data types from those used by the VM internally. I've blurred the lines with some interfaces that are used in both (because they would otherwise be identical.) What should the policy be? Personally, I would be find reusing interfaces, but I can understand the opposing view.

Reply:
What I mainly meant was that instead of doing a lot of
```typescript
    result = {
        format: disk.format,
        encoding: disk.encoding,
        volume: disk.volume,
    }
```
You can now do 
```typescript
    const { format, encoding, volume } = disk;
    const result = {
          format,
          encoding,
          volume,
    }
```
I think it's fine to reuse types for both state and interfaces, it was probably just confusing having multiple interfaces.

Reply:
The destructuring is done in all the places in `disk2.ts` where it made sense. There may be more and I'll clean those up as I find them.